Implement group backend

PVince81 opened the pull request:
Implement group manager.

To test:

1. Enable app
1. Create three users "user1", "user2", "user3"
1. `insert into oc_custom_group values (NULL, 'myfriends', 'My friends');`
1. `insert into oc_custom_group_member values (1, 'user2', 1);`
1. `insert into oc_custom_group_member values (1, 'user3', 0);`
1. Login as "user1"
1. Create a folder "test"
1. Share with "My fri" let it autocomplete: select "customgroup_1" (need future work to support display names)
1. Login as "user2" and see that the folder is visible

Some concerns:
- current impl lets any user share with anybody, unless admin ticks in admin page "only allow sharing with groups in which user is member", this could be fine as is...
- need display name support in core, might need to change some group manager APIs to support returning arrays instead of just the group ids

@DeepDiver1975 @butonic @jvillafanez what do you think ?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/customgroups?pullRequest=1) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/customgroups?pullRequest=1) before we can accept your contribution.<br/>

PVince81 commented:
also:

- custom groups are visible in user manager. Can't hide them because making `getGroups` return an empty would break searching in the sharing autocomplete. Might need changing core APIs to let the manager specify if listable or not.

jvillafanez commented on lib/CustomGroupsManager.php:
```diff
@@ -0,0 +1,196 @@
+class CustomGroupsManager implements \OCP\GroupInterface {
+
+	const GROUP_ID_PREFIX = 'customgroup_';
+	const GROUP_ID_PREFIX_LEN = 12;
```
I don't think it's a good idea to hardcode this. The extra cost of calculating the string's length should be meaningless, and we would only maintain one constant.

jvillafanez commented on lib/CustomGroupsManager.php:
```diff
@@ -0,0 +1,196 @@
+		// FIXME: search by display name instead
+		if ($search !== '') {
+			$parameters[] = '%' . $search . '%';
+			$searchLike = ' WHERE LOWER(`displayname`) LIKE LOWER(?)';
```
Do we want lowercase comparations?

PVince81 replied:
For display names I'd say yes

jvillafanez commented on lib/CustomGroupsManager.php:
```diff
@@ -0,0 +1,196 @@
+	 * @return bool
+	 */
+	public function inGroup($uid, $gid) {
+		$numericGroupId = $this->extractNumericGroupId($gid);
```
Isn't this a bit complex? Why do we want to use prefix?

jvillafanez replied:
I don't think it's the backend responsability to resolve the collisions nor to prevent them. Any other backend could have a group with the same name "customgroup_12", and there shouldn't be any reason to avoid it.

I think it's core the one who should take care of how to resolve the collisions. Generally, just saying "I want this group" is enough to know what is the group you're refering to, but maybe sometimes you should say "I want this group which is located in this backend" so there is no doubt about the group you want.

PVince81 replied:
While I agree that it's not this backend's responsibility, we'll still get into trouble if all our groups are called "1", "2", "3", etc... And I don't intend to fix core to have its own collision detection / namespacing as part of this task, which is likely a bigger task on its own which might not be backward compatible.

PVince81 commented:
- [ ] TODO: add unit tests
- [ ] TODO: remove LEN constant

jvillafanez commented:
Just a couple of extra things:

* From my point of view ( @DeepDiver1975 correct me if I'm wrong) there should be only one group manager and user manager, which are the ones provided by core. Creating and using another manager shouldn't be possible because both core and any app are interested in using the "official" manager.
* Custom backends, such as LDAP or this one could be registered in the core manager. This also includes the DB backend which is used by default in ownCloud.

PVince81 commented:
@jvillafanez yes sorry, there is only one group manager.

What this PR implements is a group backend like LDAP does. (I mixed up both names...)

PVince81 commented:
- [ ] might need adjusting the whole PR depending on how we implement the new group backend interfaces as described in https://github.com/owncloud/core/issues/7918#issuecomment-263615014 for display name support

PVince81 commented:
Anyway, my first goal with this PR was to experiment to find out whether sharing would work automatically with this approach. It did!

PVince81 commented:
- [ ] now requires https://github.com/owncloud/core/pull/26750 for display name support and the new `IGroupBackend` interface

If you test this you'll see that the display name is correctly shown in the sharee list 😄

PVince81 commented:
- [ ] TODO: every member can also be a group admin, need to change the data structure

PVince81 commented:
Ok, I spent too much time already implementing the Sabre stuff but not enough writing actual tests for GroupsManager 😄 

Will do tomorrow so we can get this first PR merged soon.

PVince81 commented:
I've added unit tests now for the group backend and group manager.
Note that for the tests to run properly you also need to checkout the matching core branch https://github.com/owncloud/core/pull/26750
(or ideally review that one first).

I've removed the DAV stuff from this PR into a separate PR: https://github.com/owncloud/customgroups/pull/7

Please review @DeepDiver1975 @jvillafanez

PVince81 commented:
@DeepDiver1975 mind reviewing so I can continue with the DAV stuff ? Thanks

jvillafanez commented on lib/CustomGroupsManager.php:
```diff
@@ -0,0 +1,321 @@
+/**
+ * Manager for custom group
+ */
+class CustomGroupsManager {
```
I'm questioning the purpose of this class... does it make the code too complex if we move this code to the backend?

Anyway, we should rename the class. The backend class will have a manager above (provided by core, where the backend will be added) and also below (this class) which is confusing. `CustomGroupsDBHandler` seems a better choice.

jvillafanez commented on appinfo/database.xml:
```diff
@@ -0,0 +1,80 @@
+				<notnull>false</notnull>
+			</field>
+			<index>
+				<name>cg_gid_uid_index</name>
```
I'm not sure if we should add another index only for the user_id. I expect the queries for groups will use this index, but I don't think the queries for users will.

jvillafanez commented on lib/Application.php:
```diff
@@ -0,0 +1,38 @@
+	 * Register the group manager
+	 */
+	public function registerGroupBackend() {
+		$backend = $this->getContainer()->query('\OCA\CustomGroups\CustomGroupsBackend');
```
I don't know where this come from.

PVince81 replied:
What do you mean ? This method is called in appinfo/app.php

jvillafanez replied:
The `\OCA\CustomGroups\CustomGroupsBackend` needs to be register in the container, but I don't know where this registration is made.

PVince81 replied:
No, it doesn't. DI container power !

It works because the class name matches exactly so it's able to resolve it.
You only need explicit registration when the names different or when your constructor requires parameters that cannot be injected (ex: strings)

jvillafanez replied:
😮 👏

PVince81 commented:
Restarted travis build now that master contains the code we need

PVince81 commented:
- [ ] rename CustomGroupsManager to something that sounds more like it's a DB accessor

PVince81 commented:
- [ ] challenge: implement initial DB structure with doctrine migration

PVince81 commented:
Looks like for the Mapper I can throw away almost all my code (and unit tests) and start over. I don't think the benefit of having the mapper pattern here is bringing any significant benefit at this point. (it might even introduce more hurdles for the API consumer)

I'll abandon this approach and simply rename the manager to `CustomGroupsDatabaseHandler`.

PVince81 commented:
Raised https://github.com/owncloud/customgroups/issues/10 to use doctrine migrations for initial DB.

I'd like to get this PR here merged so I can move forward with the Sabre plugins / API. That is, unless there is a no-go in this PR.

PVince81 commented:
Code coverage 404:
```
Uploading code coverage for repository "g/owncloud/customgroups" and revision "71b39e8571e3cba39d09c79143dabcb786640b7b"... Failed

{"error":{"code":404,"message":"Not Found"}}
```

@DeepDiver1975 do we need to create a project on that website maybe for that to work ?

PVince81 commented:
Rebased to include the doctrine migration for the initial DB, deleted database.xml.

@jvillafanez I've renamed the class to `CustomGroupsDatabaseHandler`.
I hope the PR is acceptable now. Thanks

jvillafanez commented on lib/CustomGroupsDatabaseHandler.php:
```diff
@@ -0,0 +1,321 @@
+	 * @return true if group was deleted, false otherwise
+	 */
+	public function deleteGroup($gid) {
+		// Delete the group-user relation
```
I'd wrap this in a transaction. If we want to delete a group, we want both actions to be executed or none of them. Something in the middle it be would be an unexpected result.

PVince81 commented on lib/CustomGroupsDatabaseHandler.php:
```diff
@@ -0,0 +1,321 @@
+		$cursor->closeCursor();
+
+		if (!$result) {
+			return null;
```
because `$result` is `false`, not `null`...

PVince81 commented:
Also, looks like sqlite not happy with my DB stuff. Will look into it. (Mysql worked fine)

PVince81 commented:
Something is wrong with the unique constraint. Looks like SQLite set the primary key only on "group_id" instead of both "group_id" + "user_id" 😢

PVince81 commented:
```
sqlite> pragma table_info(oc_custom_group_member);
cid|name|type|notnull|dflt_value|pk
0|group_id|INTEGER|1||1
1|user_id|VARCHAR(64)|1||0
2|is_admin|INTEGER|1|0|0
```

primary key is only set on `group_id` even though I asked it to set it on both:
```php
    $table->setPrimaryKey(['group_id', 'user_id']);
```

debugging....

PVince81 commented:
grrr, this is the resulting SQL generated by Doctrine:
```
  $sql[2]                        = (string[156]) 'CREATE TABLE oc_custom_group_member (group_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, user_id VARCHAR(64) NOT NULL, is_admin INTEGER DEFAULT 0 NOT NULL)';
```

I didn't even ask for autoincrement... Looks like a Doctrine bug.

We do have a table in core with a similar schema and that one looks correct:
```
sqlite> pragma table_info(oc_group_user);
cid|name|type|notnull|dflt_value|pk
0|gid|VARCHAR(64)|1|''|1
1|uid|VARCHAR(64)|1|''|2
```

so this means SQLite supports multi-column primary keys but Doctrine doesn't seem to be able to create that.

Googling now, I hope there is maybe a missing flag or something in the API.

PVince81 commented:
Facepalm, I forgot to remove "autoincrement" when copy-pasting the definitions.
Funnily enough, Mysql didn't complain.

PVince81 commented:
@jvillafanez I fixed all your comments and additionally fixed the DB schema.

PVince81 commented:
@DeepDiver1975 can you create the repo for clover coverage ? https://travis-ci.org/owncloud/customgroups/jobs/183218691#L831

jvillafanez commented:
👍

PVince81 commented:
I'm going to remove the code coverage thing for now and let @DeepDiver1975 sort it out separately.

PVince81 commented:
Raised here https://github.com/owncloud/customgroups/issues/12
